contents/functionalProgramming/intro.md

```diff
@@ -0,0 +1,107 @@
+
+1. [Idempotence](https://en.wikipedia.org/wiki/Idempotence) - The function should return the same output for the same input.
+1. No Side Effects - The function should not cause side effects like mutating global state.
+
+To illustrate an example of a pure function. Consider this simple example:
+
+```python
+# Pure Function
+def add(x, y):
+    return x + y
+```
+
+Contrast this with an impure function:
+
+```python
+# Mutate state
+y=0
+def addXToY(x):
+    global y
+    y = y + x
+    return y
+```
+We see that the impure function violates both conditions. Running `addXToY` multiple times with the same input X would give different value.
+
+#### Immutability
+
+Another concept in Functional Programming is immutability. When we assign `x=1`, it does not make mathematical sense to have this expression `x:=x+1`.
+
+Having immutability built into Functional Programming languages like Haskell also helps to ensure that functions are `pure` because mutable variables and states can introduce side-effects.
+
+To know more about immutability in functional languages, you can take a look at:
+
+- [Immutability in Haskell](https://mmhaskell.com/blog/2017/1/9/immutability-is-awesome)
+- [Immutability in Elm](http://elmprogramming.com/immutability.html)
+- [Why does immutability enable Functional Programming](https://stackoverflow.com/questions/12207757/why-do-immutable-objects-enable-functional-programming)
+
+## Techniques in Functional Programming
+
+#### Recursion
+
+Pure functional languages does not have loop constructs that procedural languages does. This is because `Iteration` usually involves state mutation per iteration. Since Functional Programming avoids state changes, `Recursion` is a commonly used [technique](https://www.quora.com/Why-dont-pure-functional-programming-languages-provide-a-loop-construct) to replace `Iteration`.
+
+Hence, to be able to write functional languages effectively, it means being able to replace Iteration with Recursion. Here are some guides to help you on that:
+
+- [Introduction to Recursion](https://www.topcoder.com/community/data-science/data-science-tutorials/an-introduction-to-recursion-part-1/)
+- [Recursion in Functional Programming](https://dzone.com/articles/functional-programming-recursion)
+
+#### Higher Order Functions
+
+> Higher Order Functions are functions that take functions as parameters, return functions or both.
+
+Functions are first class citizen in Functional Programming languages, this mean that functions can be passed around like objects and values. 
+
+One example of Higher Order Functions is `map`, which takes in a function and a collection of objects.
+
+```python
+length_of_names = map(len, ['Alex', 'Thomas', 'John'])
+print(length_of_names) # [4,6,4]
+```
+
+The function creates a new collection, run the function on each element of the given collection, insert the new values and return the new collection.
```

Comment:
This explanation of map is not very clear for a beginner. It doesn't even mention the name `map` so it may not be obvious to the reader. It also repeats words like `function`, `collection` many times, and each instance of the word refers to a different function/collection. I suggest something like: "`map` takes in 2 parameters, the first one is a function and the second is a collection. `map` returns a new collection whose elements are the result of applying the given function on the elements of the given collection". Even better, I think you can refer to the example you quoted, e.g. "In the example, `map` is given a function `len` and a collection `['Alex', 'Thomas', 'John']`. It returns `[len('Alex'), len('Thomas'), len('John')]`."